File: System_1/httpdocs/tools/slice_pdf.py
#!/usr/bin/env python3
import sys
import os
import json
import re
from typing import List, Dict, Optional, Any

def die(code: int, err: str, msg: str = "") -> None:
    print(json.dumps({"error": err, "message": msg}, ensure_ascii=False))
    sys.exit(code)

def norm(s: str) -> str:
    s = s.replace("\r\n", "\n").replace("\r", "\n")
    s = re.sub(r"[ \t]+", " ", s)
    s = re.sub(r"\n{3,}", "\n\n", s)
    return s.strip()

def lines(s: str) -> List[str]:
    s = s.replace("\r\n", "\n").replace("\r", "\n")
    out: List[str] = []
    for ln in s.split("\n"):
        t = ln.strip()
        if t:
            out.append(t)
    return out

def find_top1_page(page_texts: Dict[int, str], agenda_page: int, kw: List[str]) -> Dict[str, Any]:
    scanN = max(page_texts.keys()) if page_texts else 0
    hits: List[str] = []

    for p in range(min(agenda_page + 1, scanN), scanN + 1):
        ls = lines(page_texts.get(p, ""))
        n = len(ls)

        i = 0
        while i < n:
            ln = ls[i]
            looks_top1 = False

            # Spezialfall: "a. Öffentlicher Teil" -> danach nach "1.) ..." suchen
            if re.search(r"^\s*a\.\s*öffentlicher\s+teil\b", ln, re.I):
                for j in range(i + 1, min(n, i + 20)):
                    if re.search(r"^\s*1\s*[\.\)]\s+\S+", ls[j]):
                        looks_top1 = True
                        i = j
                        break

            if (not looks_top1) and re.search(r"^\s*1\s*[\.\)]\s+\S+", ln):
                looks_top1 = True
            if (not looks_top1) and re.search(r"^\s*top\s*1\b", ln, re.I):
                looks_top1 = True

            if not looks_top1:
                i += 1
                continue

            window = "\n".join(ls[i:i+35]).lower()
            for k in kw:
                kl = (k or "").strip().lower()
                if kl and kl in window:
                    hits.append(k)

            hits = sorted(set(hits))
            return {"top1_page": p, "top1_line": i + 1, "kw_hits": hits}

    return {"top1_page": None, "top1_line": None, "kw_hits": []}

def do_detect(pdf_path: str, scan_pages: int, kw: List[str]) -> Dict[str, Any]:
    try:
        from pypdf import PdfReader
    except Exception as e:
        die(1, "import_error", str(e))

    if not os.path.isfile(pdf_path):
        die(2, "src_not_found", pdf_path)

    reader = PdfReader(pdf_path)
    total = len(reader.pages)

    scanN = 0
    if total > 0:
        scanN = min(max(1, int(scan_pages)), total)

    page_texts: Dict[int, str] = {}
    agenda_page: Optional[int] = None

    for p in range(1, scanN + 1):
        try:
            txt = reader.pages[p-1].extract_text() or ""
        except Exception:
            txt = ""
        txt = norm(txt)
        page_texts[p] = txt
        if agenda_page is None and re.search(r"\btagesordnung\b", txt, re.I):
            agenda_page = p

    top = {"top1_page": None, "top1_line": None, "kw_hits": []}
    if agenda_page is not None:
        top = find_top1_page(page_texts, agenda_page, kw)

    return {
        "ok": True,
        "pages_total": total,
        "pages_scanned": scanN,
        "agenda_page": agenda_page,
        "top1_page": top["top1_page"],
        "top1_line": top["top1_line"],
        "kw_hits": top["kw_hits"],
    }

def do_slice(src: str, dst: str, end_page_1based: int) -> Dict[str, Any]:
    try:
        from pypdf import PdfReader, PdfWriter
    except Exception as e:
        die(1, "import_error", str(e))

    if end_page_1based <= 0:
        die(2, "invalid_end_page")

    if not os.path.isfile(src):
        die(2, "src_not_found", src)

    reader = PdfReader(src)
    total = len(reader.pages)
    n = min(int(end_page_1based), total)

    writer = PdfWriter()
    for i in range(n):
        writer.add_page(reader.pages[i])

    out_dir = os.path.dirname(dst)
    if out_dir and not os.path.isdir(out_dir):
        os.makedirs(out_dir, exist_ok=True)

    tmp = dst + ".tmp"
    if os.path.exists(tmp):
        try:
            os.remove(tmp)
        except Exception:
            pass

    with open(tmp, "wb") as f:
        writer.write(f)

    os.replace(tmp, dst)

    return {"ok": True, "pages_total": total, "pages_written": n, "dst": dst}

def main() -> None:
    # Neu: detect
    #   slice_pdf.py detect --pdf <pdf> [--scan_pages N] [--kw <k>]...
    # Alt: slice
    #   slice_pdf.py <src_pdf> <dst_pdf> <end_page_1based>
    if len(sys.argv) >= 2 and sys.argv[1] == "detect":
        pdf: Optional[str] = None
        scan_pages = 25
        kw: List[str] = []

        i = 2
        while i < len(sys.argv):
            a = sys.argv[i]
            if a == "--pdf" and i + 1 < len(sys.argv):
                pdf = sys.argv[i + 1]
                i += 2
                continue
            if a == "--scan_pages" and i + 1 < len(sys.argv):
                try:
                    scan_pages = int(sys.argv[i + 1])
                except Exception:
                    die(2, "invalid_scan_pages")
                i += 2
                continue
            if a == "--kw" and i + 1 < len(sys.argv):
                kw.append(sys.argv[i + 1])
                i += 2
                continue

            die(2, "bad_args", f"unknown or incomplete arg: {a}")

        if not pdf:
            die(2, "missing_args", "usage: slice_pdf.py detect --pdf <pdf> [--scan_pages N] [--kw <k>]...")

        try:
            res = do_detect(pdf, scan_pages, kw)
            print(json.dumps(res, ensure_ascii=False))
            sys.exit(0)
        except SystemExit:
            raise
        except Exception as e:
            die(1, "exception", str(e))

    # Fallback: alter Slice-Modus (abwärtskompatibel)
    if len(sys.argv) < 4:
        die(2, "missing_args", "usage: slice_pdf.py <src_pdf> <dst_pdf> <end_page_1based>")

    src = sys.argv[1]
    dst = sys.argv[2]
    try:
        end_page = int(sys.argv[3])
    except Exception:
        die(2, "invalid_end_page")

    try:
        res = do_slice(src, dst, end_page)
        print(json.dumps(res, ensure_ascii=False))
        sys.exit(0)
    except SystemExit:
        raise
    except Exception as e:
        die(1, "exception", str(e))

if __name__ == "__main__":
    main()

File: System_1/httpdocs/tools/slice_pdf.php
<?php
declare(strict_types=1);

header('Content-Type: text/plain; charset=utf-8');

/*
  slice_pdf.php
  CLI: php tools/slice_pdf.php --doc_id=123

  Output:
    <docId>_slice.pdf
    <docId>_slice.meta.json
  direkt neben dem Original PDF
*/

require __DIR__ . '/../src/db.php';
$cfg = require __DIR__ . '/../src/config.php';
require __DIR__ . '/../vendor/autoload.php';

/* ===================== helpers ===================== */

function arg_doc_id(array $argv): int {
  foreach ($argv as $a) {
    if (preg_match('/^--doc_id=(\d+)$/', (string)$a, $m)) return (int)$m[1];
  }
  return 0;
}

function abs_pdf_path(array $cfg, string $storagePath): string {
  $baseDir = rtrim((string)$cfg['storage']['base_dir'], '/');
  return $baseDir . '/' . ltrim($storagePath, '/');
}

function ensure_dir(string $dir): void {
  if (is_dir($dir)) return;
  if (!@mkdir($dir, 0775, true) && !is_dir($dir)) {
    throw new RuntimeException('Konnte Zielordner nicht erstellen: ' . $dir);
  }
}

function run_python(
  string $pythonBin,
  string $scriptPath,
  array $args,
  ?string $pyPkgsDir
): array {

  if (!is_file($scriptPath)) {
    throw new RuntimeException("Python Script fehlt: $scriptPath");
  }

  $env = $_ENV;
  if ($pyPkgsDir !== null && is_dir($pyPkgsDir)) {
    $env['PYTHONPATH'] =
      $pyPkgsDir .
      (isset($env['PYTHONPATH']) ? (PATH_SEPARATOR . $env['PYTHONPATH']) : '');
  }

  $cmd = $pythonBin . ' ' . escapeshellarg($scriptPath);
  foreach ($args as $a) {
    $cmd .= ' ' . escapeshellarg((string)$a);
  }

  $descs = [
    0 => ['pipe', 'r'],
    1 => ['pipe', 'w'],
    2 => ['pipe', 'w'],
  ];

  $proc = @proc_open($cmd, $descs, $pipes, __DIR__ . '/..', $env);
  if (!is_resource($proc)) {
    throw new RuntimeException("proc_open fehlgeschlagen: $cmd");
  }

  fclose($pipes[0]);
  $stdout = stream_get_contents($pipes[1]) ?: '';
  $stderr = stream_get_contents($pipes[2]) ?: '';
  fclose($pipes[1]);
  fclose($pipes[2]);
  $code = proc_close($proc);

  $raw = trim($stdout);
  $data = json_decode($raw, true);

  if (!is_array($data)) {
    $p = strpos($raw, '{');
    if ($p !== false) {
      $data = json_decode(substr($raw, $p), true);
    }
  }

  if (!is_array($data)) {
    $hint = mb_substr($raw, 0, 800, 'UTF-8');
    $errHint = trim($stderr) !== '' ? ("\nPY STDERR:\n" . mb_substr($stderr, 0, 800, 'UTF-8')) : '';
    throw new RuntimeException("Ungültiges JSON aus Python.\n$hint$errHint");
  }

  if (isset($data['error'])) {
    $msg = (string)($data['message'] ?? '');
    throw new RuntimeException("Python error=" . (string)$data['error'] . ($msg !== '' ? (" msg=" . $msg) : ''));
  }

  if ($code !== 0) {
    $msg = trim($stderr) !== '' ? trim($stderr) : "Python exit=$code";
    throw new RuntimeException($msg);
  }

  return $data;
}

/* ===================== main ===================== */

$docId = arg_doc_id($argv);
if ($docId <= 0) {
  fwrite(STDERR, "Fehlendes --doc_id\n");
  exit(2);
}

$baseDir = rtrim((string)($cfg['storage']['base_dir'] ?? ''), '/');
if ($baseDir === '') {
  fwrite(STDERR, "storage.base_dir fehlt\n");
  exit(2);
}

try {

  $pdo = db();
  $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

  $st = $pdo->prepare("SELECT storage_path FROM documents WHERE id=? AND deleted_at IS NULL");
  $st->execute([$docId]);
  $rel = (string)$st->fetchColumn();

  if ($rel === '' || $rel === 'PENDING') {
    throw new RuntimeException('storage_path fehlt für doc_id=' . $docId);
  }

  $srcAbs = abs_pdf_path($cfg, $rel);
  if (!is_file($srcAbs)) {
    throw new RuntimeException('PDF nicht gefunden: ' . $srcAbs);
  }

  $dir = dirname($srcAbs);
  $dstAbs  = $dir . '/' . $docId . '_slice.pdf';
  $metaAbs = $dir . '/' . $docId . '_slice.meta.json';

  if (is_file($dstAbs)) {
    echo "Slice existiert bereits\n";
    exit(0);
  }

  $kw = [
    'bericht erstattet',
    'beschließt',
    'beschliesst',
    'nimmt kenntnis',
    'die bezirksvertretung',
    'der rat',
    'der ausschuss',
    'abstimmung',
  ];

  $pythonBin = is_file(__DIR__ . '/../.venv/bin/python3')
    ? (__DIR__ . '/../.venv/bin/python3')
    : 'python3';

  $pyScript  = __DIR__ . '/slice_pdf.py';
  $pyPkgsDir = is_dir(__DIR__ . '/../py_pkgs') ? (__DIR__ . '/../py_pkgs') : null;

  $scanN = 25;

  /* ---------- 1) detect ---------- */

  $detectArgs = ['detect', '--pdf', $srcAbs, '--scan_pages', (string)$scanN];
  foreach ($kw as $k) {
    $detectArgs[] = '--kw';
    $detectArgs[] = $k;
  }

  $detect = run_python($pythonBin, $pyScript, $detectArgs, $pyPkgsDir);

  $total = (int)($detect['pages_total'] ?? 0);
  if ($total <= 0) {
    throw new RuntimeException('detect lieferte pages_total=0');
  }

  $agendaPage = $detect['agenda_page'] ?? null;
  $top1Page   = $detect['top1_page'] ?? null;
  $top1Line   = $detect['top1_line'] ?? null;
  $kwHits     = $detect['kw_hits'] ?? [];

  /* ---------- 2) sliceEnd berechnen ---------- */

  $sliceEnd = null;
  $contaminated = false;

  if ($agendaPage !== null && $top1Page !== null) {

    $top1PageI = (int)$top1Page;
    $top1LineI = (int)($top1Line ?? 0);

    if ($top1LineI > 0 && $top1LineI <= 15) {
      $sliceEnd = $top1PageI - 1;
    } else {
      $sliceEnd = $top1PageI;
      $contaminated = true;
    }

  } else {
    $sliceEnd = min($total, 12);
    $contaminated = true;
  }

  $sliceEnd = max(1, min((int)$sliceEnd, $total));

  /* ---------- 3) slice ---------- */

  ensure_dir(dirname($dstAbs));

  $slice = run_python(
    $pythonBin,
    $pyScript,
    [$srcAbs, $dstAbs, (string)$sliceEnd],   // alter Slice-Modus bleibt kompatibel
    $pyPkgsDir
  );

  /* ---------- 4) meta ---------- */

  $meta = [
    'doc_id' => $docId,
    'source_pdf' => $srcAbs,
    'slice_pdf' => $dstAbs,
    'created_at' => date('c'),
    'pages_total' => $total,
    'pages_scanned' => min($total, $scanN),
    'agenda_page' => $agendaPage,
    'top1_page' => $top1Page,
    'top1_line_index' => $top1Line,
    'top1_keyword_hits' => $kwHits,
    'slice_end_page' => $sliceEnd,
    'contaminated' => $contaminated,
    'python_detect' => $detect,
    'python_slice' => $slice,
    'notes' => 'detect + slice via python (pypdf). no smalot/pdfparser.',
  ];

  @file_put_contents(
    $metaAbs,
    json_encode($meta, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE)
  );

  echo "OK slice_end_page=" . $sliceEnd . " contaminated=" . ($contaminated ? '1' : '0') . "\n";
  exit(0);

} catch (Throwable $e) {
  fwrite(STDERR, $e->getMessage() . "\n");
  exit(1);
}
